### PR TITLE
SDIT-1490 more refactoring

### PR DESCRIPTION
**Pull request by mikehalmamoj**

### CODE REVIEW

File: hmpps-prisoner-search/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/api/Matchers.kt [hunk 1]
```diff
@@ -4,41 +4,31 @@ import uk.gov.justice.digital.hmpps.prisonersearch.search.services.attributesear
 
 data class Matchers(
   val joinType: JoinType,
-  val stringMatchers: List<StringMatcher>? = null,
-  val booleanMatchers: List<BooleanMatcher>? = null,
-  val intMatchers: List<IntMatcher>? = null,
-  val dateMatchers: List<DateMatcher>? = null,
-  val dateTimeMatchers: List<DateTimeMatcher>? = null,
```
Comment (mikehalmamoj) [inline hmpps-prisoner-search/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/api/Matchers.kt:11]: The main change is to consolidate the type matchers into a single list

File: hmpps-prisoner-search/src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/dto/AttributeSearchRequestJsonTest.kt [hunk 2]
```diff
@@ -296,14 +307,12 @@ class AttributeSearchRequestJsonTest(@Autowired val objectMapper: ObjectMapper)
...
             joinType = JoinType.AND,
-            stringMatchers = listOf(
+            matchers = listOf(
               StringMatcher(
                 attribute = "firstName",
                 condition = TextCondition.IS,
                 searchTerm = "John",
               ),
-            ),
-            booleanMatchers = listOf(
```
Comment (mikehalmamoj) [inline hmpps-prisoner-search/src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/dto/AttributeSearchRequestJsonTest.kt:306]: With a single list of matchers we don't discrimate by matcher type which felt a bit clunky

File: hmpps-prisoner-search/src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/dto/AttributeSearchRequestJsonTest.kt [hunk 3]
```diff
@@ -272,15 +283,15 @@ class AttributeSearchRequestJsonTest(@Autowired val objectMapper: ObjectMapper)
...
                      "searchTerm": "John"
-                   }
-                 ],
-                 "booleanMatchers": [
+                   },
                    {
+                     "type": "Boolean",
                      "attribute": "recall",
                      "condition": true
                    }
```
Comment (mikehalmamoj) [inline hmpps-prisoner-search/src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/dto/AttributeSearchRequestJsonTest.kt:297]: With a single list of type matchers we don't have separate lists for each type which felt a bit clunky. Instead each matcher has a `type` discrimnator.

File: hmpps-prisoner-search/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/AtttributeResolver.kt [hunk 4]
```diff
@@ -17,31 +17,24 @@ class AttributeResolver {
...
       getGenericTypeClass(prop).memberProperties
-        .forEach { childProp -> findTypes(childProp, map, "${prefix}${prop.name}.") }
+        .flatMap { childProp -> findTypes(childProp, "${prefix}${prop.name}.") }
     }
     PropertyType.COMPLEX -> {
       getPropertyClass(prop).memberProperties
-        .forEach { childProp -> findTypes(childProp, map, "${prefix}${prop.name}.") }
+        .flatMap { childProp -> findTypes(childProp, "${prefix}${prop.name}.") }
     }
   }
```
Comment (mikehalmamoj) [inline hmpps-prisoner-search/src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/AtttributeResolver.kt:37]: Remvoe the mutable list and forEach to make this a bit slicker

File: hmpps-prisoner-search/src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/AttributeResolverTest.kt [hunk 5]
```diff
@@ -39,26 +40,28 @@ class AttributeResolverTest {
...
-      of("list.nested.description", String::class),
-      of("complex.string", String::class),
-      of("complex.boolean", Boolean::class),
-      of("complex.integer", Integer::class),
-      of("complex.localDate", LocalDate::class),
-      of("complex.localDateTime", LocalDateTime::class),
-      of("complex.nested.code", String::class),
-      of("complex.nested.description", String::class),
+    fun testParameters(): Stream<Arguments> = Stream.of(
+      arguments(
```
Comment (petergphillips) [inline hmpps-prisoner-search/src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/AttributeResolverTest.kt:44]: apologies I more meant just replacing each `of` with `arguments` i.e.
```kotlin
      arguments("string", String::class),
      arguments("boolean", Boolean::class),
      arguments("integer", Integer::class),
```

**Review by petergphillips — APPROVED** [review]
LGTM

File: hmpps-prisoner-search/src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/AttributeResolverTest.kt (hunk 5 above)
Comment (mikehalmamoj) [reply hmpps-prisoner-search/src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/AttributeResolverTest.kt:44]: I quite like the vararg version of arguments tbh - even less duplication

Comment (mikehalmamoj) [reply hmpps-prisoner-search/src/test/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/search/services/attributesearch/AttributeResolverTest.kt:44]: Oh except it doesn't work 😆  Whoops.

**Pull request merged**